telemetry/metrics/init.go

```diff
@@ -0,0 +1,117 @@
+package metrics
+
+import (
+	"context"
+	"math/rand"
+
+	"github.com/gnolang/gno/telemetry/exporter"
+	"github.com/gnolang/gno/telemetry/options"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/metric"
+	sdkMetric "go.opentelemetry.io/otel/sdk/metric"
+)
+
+var (
+	ctx context.Context
+
+	// Metrics.
+	BroadcastTxTimer Int64Collector
+	BuildBlockTimer  Int64Collector
+)
+
+func Init(setCtx context.Context, config options.Config) error {
+	if config.ExporterEndpoint == "" {
+		return exporter.ErrEndpointNotSet
+	}
+
+	ctx = setCtx
+
+	// Use oltp metric exporter.
+	exporter, err := otlpmetricgrpc.New(
+		ctx,
+		otlpmetricgrpc.WithEndpoint(config.ExporterEndpoint),
+		otlpmetricgrpc.WithInsecure(), // TODO: enable security
+	)
+	if err != nil {
+		return err
+	}
+
+	provider := sdkMetric.NewMeterProvider(sdkMetric.WithReader(sdkMetric.NewPeriodicReader(exporter)))
+	otel.SetMeterProvider(provider)
+	meter := provider.Meter(
+		config.MeterName,
+		metric.WithInstrumentationVersion("0.1.0"),
+		metric.WithSchemaURL(config.ServiceName),
+	)
+
+	broadcastTxTimer, err := meter.Int64Histogram(
+		"broadcast_tx_hist",
+		metric.WithDescription("broadcast tx duration"),
+		// metric.WithExplicitBucketBoundaries(0, 16, 32, 64, 128, 256, 512),
+		metric.WithUnit("ms"),
+	)
+	if err != nil {
+		return err
+	}
+	BroadcastTxTimer = Int64Histogram{
+		Int64Histogram: broadcastTxTimer,
+		useFakeMetrics: config.UseFakeMetrics,
+		fakeRangeStart: 5,
+		fakeRangeEnd:   250,
+	}
+
+	buildBlockTimer, err := meter.Int64Histogram(
+		"build_block_hist",
+		metric.WithDescription("block build duration"),
+		// metric.WithExplicitBucketBoundaries(0, 16, 32, 64, 128, 256, 512),
+		metric.WithUnit("ms"),
+	)
+	if err != nil {
+		return err
+	}
+	BuildBlockTimer = Int64Histogram{
+		Int64Histogram: buildBlockTimer,
+		useFakeMetrics: config.UseFakeMetrics,
+		fakeRangeStart: 0,
+		fakeRangeEnd:   150,
+	}
+
+	return nil
+}
+
+type Int64Collector interface {
+	Collect(int64)
+}
+
```

Comment:
I think you can completely drop this interface, and use the exported types (that implement this)

Usually modules will:
- define their own metrics client API (as an interface, preferred)
- use an already given API (enforced by the package, less preferred)

With the first approach we can mock module-level if we wanted to test the metrics individually, and is more flexible for the module if our needs change

Reply:
Dropped